### PR TITLE
DHIS2-11206 Custom Enrollment Form

### PR DESCRIPTION
**Pull request by EirikHaugstulen**
This PR introduces compatibility with the Custom Enrollment Form that can be set up in the maintenance app.
The custom form takes precedence over both the basic form and section forms. If none are present, Capture will revert back to the basic form. The custom form will also work with styling and script-tags.

As of now, there are no support for custom program enrollment details (incident date, enrollment date) or tracked entity feature type. This will be rendered as their own section above the custom form.

This PR builds on a refactoring of the `EnrollmentFactory.js` which was introduced in [Sections in Enrollment Form-PR](https://github.com/dhis2/capture-app/pull/1608). This will have to be merged before this issue.
Thanks!

### CODE REVIEW

**Comment by simonadomnisoru** [comment]
Hey @EirikHaugstulen, 

Do you plan to add some Cypress Scenarios to verify the new custom enrollment form functionality? Thanks!

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            };
+
+            return React.createElement(
+                'FormField', {
+                    customFormElementProps,
+                    id: htmlElementId,
+                },
+            );
+        }
+        return null;
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js:27]: Do we need this? To be clear: If we have an input node without an attributeId I would prefer that it was rendered to the DOM without any transformation.

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js [hunk 2]
```diff
@@ -34,13 +34,17 @@ export class CustomForm {
      *
      * @memberof CustomForm
      */
-    set data(html: string) {
-        const data = parseHtml(html, {
-            onTransform: this.transformNode,
+    setData(html: string, transformFunction: () => ?React.Element<'FormField'>) {
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js:37]: NIT: I think it would be a good exercise to improve up on this Flow declaration. (Part 1)

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js [hunk 3]
```diff
@@ -91,63 +85,138 @@ export class EnrollmentFactory {
...
+        const trackedEntityTypeName = this.cachedTrackedEntityTypes.get(cachedProgramTrackedEntityTypeId);
+
+        if (!trackedEntityTypeName?.displayName || !featureTypeField) {
             return null;
         }
 
+        const section = new Section((o) => {
+            o.id = cachedProgramTrackedEntityTypeId;
+            o.name = trackedEntityTypeName.displayName;
+        });
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js:101]: I'm going to nitpick a bit here 😇 
- `trackedEntityTypeName` is really `trackedEntityType`
- If for some reason the `trackedEntityType` is not found or doesn't have a `displayName`, I think we should still  show the section (without a name).

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js [hunk 4]
```diff
@@ -34,13 +34,17 @@ export class CustomForm {
...
         });
+    }
+
+    set data(data: Object) {
         this._data = data;
     }
+
     get data(): Data {
         return this._data;
     }
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js:50]: I think we can safely remove these?

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js [hunk 5]
```diff
@@ -91,63 +85,138 @@ export class EnrollmentFactory {
...
+        section.showContainer = false;
+
+        section = await this._buildElementsForSection(cachedProgramTrackedEntityAttributes, section);
+        section && enrollmentForm.addSection(section);
+        try {
+            enrollmentForm.customForm = new CustomForm((o) => {
+                o.id = dataEntryForm.id;
+            });
+            // $FlowFixMe : Require input from class
+            enrollmentForm.customForm.setData(dataEntryForm.htmlCode, transformTrackerNode);
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js:177]: NIT: Flow declaration part 2

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+// @flow
+import React from 'react';
+
+export const transformTrackerNode = (node: Object, index: number, nodeToElementFn: (...args : Array<any>) => any) => {
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js:4]: NIT: Flow declaration part3. `nodeToElementFn` can be more specific I think. We can have a chat about these flow declarations, but I think it's a good exercise to look into this a bit.

File: src/core_modules/capture-core/components/D2Form/D2Form.component.js [hunk 7]
```diff
@@ -110,9 +110,11 @@ class D2Form extends React.PureComponent<PropsForPureComponent> {
             key={section.id}
         >
             <D2SectionContainer
-                innerRef={(sectionInstance) => { this.setSectionInstance(sectionInstance, section.id); }}
+                innerRef={(sectionInstance) => {
+                    this.setSectionInstance(sectionInstance, section.id);
+                }}
                 sectionMetaData={section}
-                customForm={this.props.formFoundation.customForm}
+                customForm={section.id === '#MAIN#' ? this.props.formFoundation.customForm : null}
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/D2Form/D2Form.component.js:117]: Do you think that skipping the ternary operator here will be fine? 
```suggestion
                customForm={this.props.formFoundation.customForm}
```

**Review by JoakimSM — COMMENTED** [review]
Looks really good @EirikHaugstulen 👍 The comments are me mostly nitpicking.

File: src/core_modules/capture-core/components/D2Form/D2Form.component.js (hunk 7 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/components/D2Form/D2Form.component.js:117]: The ternary operator is there in case the enrollment form has a `Tracked Entity Feature Type`. If there is a Feature type, the form will render a separate section. If we don't use the ternary operator, the form will change the html of all available sections in the formFoundation, including the section for Feature Type. (And any other sections if we ever have to implement it in the future). 
This way, it will only replace the HTML in the `#MAIN#`-section.
Let me know if this is unclear or if I should still make any changes.

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js (hunk 4 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js:50]: The program will throw an error if we remove the `get data()`, but no errors on removing `set Data()`.
Is it okay if we keep `get data()`, or should I try to find a fix?

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        return null;
+    }
+    return undefined;
+};
+
+export const transformEventNode = (node: Object, index: number, nodeToElementFn: (...args : Array<any>) => any) => {
+    if (node.name === 'input') {
+        const htmlElementId = node.attribs && node.attribs.id;
+        const matchResult = htmlElementId && /-[^-]+/.exec(htmlElementId);
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js:35]: Hey, 
I think that having the `/-[^-]+/`regex stored in a const with a descriptive name can improve the code readability. WDYT? Thanks!

File: src/core_modules/capture-core/components/D2Form/field/configs/base/configBaseCustomForm.js [hunk 9]
```diff
@@ -4,8 +4,9 @@ import { getBaseConfigForField } from './configBase';
 import type { DataElement as MetaDataElement } from '../../../../../metaData';
 import type { FieldConfigForType } from './configBase';
 
-const getBaseProps = (metaData: MetaDataElement) => ({
-    metaCompulsory: metaData.compulsory,
+const getBaseProps = ({ compulsory, disabled, unique }: MetaDataElement) => ({
+    metaCompulsory: compulsory,
+    metaDisabled: disabled || (unique && unique.generatable),
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/D2Form/field/configs/base/configBaseCustomForm.js:9]: NIT: You can use optional chaining: `metaDisabled: disabled || unique?.generatable,`

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js [hunk 10]
```diff
@@ -34,13 +34,17 @@ export class CustomForm {
...
-        const data = parseHtml(html, {
-            onTransform: this.transformNode,
+    setData(html: string, transformFunction: () => ?React.Element<'FormField'>) {
+        this._data = parseHtml(html, {
+            onTransform: transformFunction,
             allowScript: true,
         });
+    }
+
+    set data(data: Object) {
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js:44]: Can you help me understand why `data` is not of type `Data` here? In both the initialisation and the getter type `Data` is used

File: src/core_modules/capture-core/components/D2Form/D2Form.component.js (hunk 7 above)
Comment (JoakimSM) [reply src/core_modules/capture-core/components/D2Form/D2Form.component.js:117]: Ah, I see. Follow-up suggestion: Can customForm be a property of the Section object instead of the RenderFoundation object?

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js (hunk 4 above)
Comment (JoakimSM) [reply src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js:50]: Good point 👍 Yep, let's keep `get data()` and remove the setter.

File: src/core_modules/capture-core/metaDataStoreLoaders/programs/quickStoreOperations/storePrograms.js [hunk 11]
```diff
@@ -27,6 +27,15 @@ const convert = (() => {
                 return accTranslationObject;
             }, {});
 
+    const convertProgramSections = apiProgramSections =>
+        (apiProgramSections ? apiProgramSections
+            .filter(apiProgramSection => apiProgramSection.displayFormName)
```
[the remark this answers is not in the document]
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaDataStoreLoaders/programs/quickStoreOperations/storePrograms.js]: Sorry, don't know what I was thinking. Removed.

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js (hunk 5 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js]: I think I've been more specific on the flow declarations, but having some trouble on what to change on this one. The transformTrackerNode takes node and index as input and they are both sent in from the `CustomForm`-class. Therefore, this callback-function will not have any input even though flow expects there to be. Could you give me a hint of what to do? Also be sure to let me know if you still think the other declarations could be improved. Thanks!

File: src/core_modules/capture-core/components/D2Form/D2Form.component.js (hunk 7 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/components/D2Form/D2Form.component.js]: Good idea 👍 Implemented!

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js (hunk 8 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js]: Hey,
Absolutely agree, implemented!
Thanks

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js (hunk 10 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js]: Sorry, this one slipped through my filters. The whole set-method is removed as we already have a function to set the data. Thanks for the feedback!

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js (hunk 1 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/transformNodeFuntions/transformNodeFunctions.js]: Noted, I've removed it.

**Comment by EirikHaugstulen** [comment]
> Hey @EirikHaugstulen,
> 
> Do you plan to add some Cypress Scenarios to verify the new custom enrollment form functionality? Thanks!

Hey @simonadomnisoru,
Sorry for the late reply, but needed to have a chat with Joakim to find the best approach for testing.

Since we ideally want to keep the demo database as clean as possible, with realistic data, there are two possible outcomes to the testing:

1.  Write a cypress test that post the customForm to the api, runs the tests, and then sends another request to clean up the server data.
2. Use a demo-database that won't have as "pure" data, but will be free for all devs to implement changes, even if they are "edge-case tests". (As Joakim have mentioned previously) 

If we eventually end up going for solution 2, the first one will take unnecessary long time and would require a rewrite either way. 

I think that we need to decide this issue before writing tests that will be very slow performant.
Be sure to correct me if I am wrong @JoakimSM!
Thanks!

**Review by simonadomnisoru — APPROVED** [review]
LGTM!

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js [hunk 12]
```diff
@@ -26,56 +26,22 @@ export class CustomForm {
...
     }
     /**
      * parses html to react elements
      *
      * @memberof CustomForm
      */
-    set data(html: string) {
-        const data = parseHtml(html, {
-            onTransform: this.transformNode,
+    setData(html: string, transformFunction: (node: Object, index: number) => ?Element<'FormField'>) {
```
Comment (JoakimSM) [inline src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js:37]:
```suggestion
   setData(html: string, transformFunction: TransformFunction) {
```

and the type definition: `type TransformFunction =
    (node: Object, index: number, nodeToElementFn: (node: Object, index: number) => Element<'FormField'>) =>
        void | Element<'FormField'>;` (it's ok to put this in the CustomForm.js at the top).

Should fix the Flow error you have in the EnrollmentFactory (line 176)

**Review by JoakimSM — COMMENTED** [review]
Thanks, looks good. Made a suggestion for the Flow error.

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js (hunk 5 above)
Comment (JoakimSM) [reply src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js]: Made a suggestion for this (see other comment)

File: src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js (hunk 12 above)
Comment (EirikHaugstulen) [reply src/core_modules/capture-core/metaData/RenderFoundation/CustomForm.js]: Thanks, @JoakimSM! 👍 Implemented.

**Review by huocha — APPROVED** [review]

**Review by JoakimSM — APPROVED** [review]
